In [1]:
import pandas as pd
import numpy as np

train1 = pd.read_csv('/Users/leidelong/competition/Porto_Seguro_Safe_Driver_Prediction/input/train.csv')
# test1 = pd.read_csv("/Users/leidelong/competition/Porto_Seguro_Safe_Driver_Prediction/input/test.csv")

train2 = pd.read_csv('/Users/leidelong/competition/Porto_Seguro_Safe_Driver_Prediction/input/train_extends.csv')

target='target'
IDcol = 'id'
tag='tag'


In [2]:
print train2.columns

Index([u'Unnamed: 0', u'id', u'ps_calc_01', u'ps_calc_02', u'ps_calc_03',
       u'ps_calc_04', u'ps_calc_05', u'ps_calc_06', u'ps_calc_07',
       u'ps_calc_08',
       ...
       u'ps_car_15^2', u'ps_car_15 ps_calc_01', u'ps_car_15 ps_calc_02',
       u'ps_car_15 ps_calc_03', u'ps_calc_01^2', u'ps_calc_01 ps_calc_02',
       u'ps_calc_01 ps_calc_03', u'ps_calc_02^2', u'ps_calc_02 ps_calc_03',
       u'ps_calc_03^2'],
      dtype='object', length=176)


In [3]:
#拆分数据集
from sklearn.cross_validation import train_test_split  

#标注正样本和负样本
train1_pos=train1[train1['target']==1]
print (train1_pos['target'].value_counts())
train1_neg=train1[train1['target']==0]
print (train1_neg['target'].value_counts())

dtypes = train1.dtypes
cols = train1.columns
ty = dict(zip(cols,dtypes))

#df转narray
train1_pos=np.array(train1_pos)
train1_neg=np.array(train1_neg)
#数据集拆分
train1_pos_new,validation1_pos_new = train_test_split(train1_pos,  
                                                   test_size = 0.2,  
                                                   random_state = 0)  
train1_neg_new,validation1_neg_new = train_test_split(train1_neg,  
                                                   test_size = 0.2,  
                                                   random_state = 0)  
#数据集拼接
train1_new=np.concatenate([train1_pos_new,train1_neg_new],axis=0) #在纵轴上合并

validation1_new=np.concatenate([validation1_pos_new,validation1_neg_new],axis=0) #在纵轴上合并


#修改train、validation、test
train_temp=pd.DataFrame(train1_new)
train_temp.columns=train1.columns
train=train_temp
train_old = train.astype(ty)

validation_temp=pd.DataFrame(validation1_new)
validation_temp.columns=train1.columns
validation=validation_temp
validation_old = validation.astype(ty)


/Library/Python/2.7/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


1    21694
Name: target, dtype: int64
0    573518
Name: target, dtype: int64


In [4]:
#拆分数据集
from sklearn.cross_validation import train_test_split  

#标注正样本和负样本
train2_pos=train2[train2['target']==1]
print (train2_pos['target'].value_counts())
train2_neg=train2[train2['target']==0]
print (train2_neg['target'].value_counts())

dtypes = train2.dtypes
cols = train2.columns
ty = dict(zip(cols,dtypes))

#df转narray
train2_pos=np.array(train2_pos)
train2_neg=np.array(train2_neg)
#数据集拆分
train2_pos_new,validation2_pos_new = train_test_split(train2_pos,  
                                                   test_size = 0.2,  
                                                   random_state = 0)  
train2_neg_new,validation2_neg_new = train_test_split(train2_neg,  
                                                   test_size = 0.2,  
                                                   random_state = 0)  
#数据集拼接
train2_new=np.concatenate([train2_pos_new,train2_neg_new],axis=0) #在纵轴上合并

validation2_new=np.concatenate([validation2_pos_new,validation2_neg_new],axis=0) #在纵轴上合并


#修改train、validation、test
train_temp=pd.DataFrame(train2_new)
train_temp.columns=train2.columns
train=train_temp
train_extends = train.astype(ty)

validation_temp=pd.DataFrame(validation2_new)
validation_temp.columns=train2.columns
validation=validation_temp
validation_extends = validation.astype(ty)

1    21694
Name: target, dtype: int64
0    573518
Name: target, dtype: int64


In [5]:
print train_old.shape
print train_old.dtypes

(476169, 59)
id                  int64
target              int64
ps_ind_01           int64
ps_ind_02_cat       int64
ps_ind_03           int64
ps_ind_04_cat       int64
ps_ind_05_cat       int64
ps_ind_06_bin       int64
ps_ind_07_bin       int64
ps_ind_08_bin       int64
ps_ind_09_bin       int64
ps_ind_10_bin       int64
ps_ind_11_bin       int64
ps_ind_12_bin       int64
ps_ind_13_bin       int64
ps_ind_14           int64
ps_ind_15           int64
ps_ind_16_bin       int64
ps_ind_17_bin       int64
ps_ind_18_bin       int64
ps_reg_01         float64
ps_reg_02         float64
ps_reg_03         float64
ps_car_01_cat       int64
ps_car_02_cat       int64
ps_car_03_cat       int64
ps_car_04_cat       int64
ps_car_05_cat       int64
ps_car_06_cat       int64
ps_car_07_cat       int64
ps_car_08_cat       int64
ps_car_09_cat       int64
ps_car_10_cat       int64
ps_car_11_cat       int64
ps_car_11           int64
ps_car_12         float64
ps_car_13         float64
ps_car_14         float64

In [6]:
print train_extends.shape
print train_extends.dtypes


(476169, 176)
Unnamed: 0                 int64
id                         int64
ps_calc_01               float64
ps_calc_02               float64
ps_calc_03               float64
ps_calc_04                 int64
ps_calc_05                 int64
ps_calc_06                 int64
ps_calc_07                 int64
ps_calc_08                 int64
ps_calc_09                 int64
ps_calc_10                 int64
ps_calc_11                 int64
ps_calc_12                 int64
ps_calc_13                 int64
ps_calc_14                 int64
ps_calc_15_bin             int64
ps_calc_16_bin             int64
ps_calc_17_bin             int64
ps_calc_18_bin             int64
ps_calc_19_bin             int64
ps_calc_20_bin             int64
ps_car_11                float64
ps_car_12                float64
ps_car_13                float64
ps_car_14                float64
ps_car_15                float64
ps_ind_01                  int64
ps_ind_03                  int64
ps_ind_06_bin              in

In [7]:
#模型
from sklearn.ensemble import GradientBoostingClassifier
import xgboost as xgb
from xgboost.sklearn import XGBClassifier


from sklearn import cross_validation, metrics
from sklearn.grid_search import GridSearchCV

import matplotlib.pylab as plt
%matplotlib inline
from matplotlib.pylab import rcParams
rcParams['figure.figsize'] = 12, 4


/Library/Python/2.7/site-packages/sklearn/grid_search.py:42: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.
  DeprecationWarning)


In [8]:
#评估方法
def gini(actual, pred, cmpcol = 0, sortcol = 1):
     assert( len(actual) == len(pred) )
     all = np.asarray(np.c_[ actual, pred, np.arange(len(actual)) ], dtype=np.float)
     all = all[ np.lexsort((all[:,2], -1*all[:,1])) ]
     totalLosses = all[:,0].sum()
     giniSum = all[:,0].cumsum().sum() / totalLosses
 
     giniSum -= (len(actual) + 1) / 2.
     return giniSum / len(actual)

def gini_normalized(a, p):
        return gini(a, p) / gini(a, a)
    


#模型训练
def xgboostModelFit(alg, dtrain, dtest, predictors,useTrainCV=True, cv_folds=5, early_stopping_rounds=50):
    
    if useTrainCV:
        xgb_param = alg.get_xgb_params()
        xgtrain = xgb.DMatrix(dtrain[predictors].values, label=dtrain[target].values)
        xgtest = xgb.DMatrix(dtest[predictors].values)
        cvresult = xgb.cv(xgb_param, xgtrain, num_boost_round=alg.get_params()['n_estimators'], nfold=cv_folds,
            metrics='auc', early_stopping_rounds=early_stopping_rounds)
        alg.set_params(n_estimators=cvresult.shape[0])
    
    #模型训练
    alg.fit(dtrain[predictors], dtrain['target'],eval_metric='auc')
        
    #模型预测-训练集
    dtrain_predictions = alg.predict(dtrain[predictors])
    dtrain_predprob = alg.predict_proba(dtrain[predictors])[:,1]
        
    #输出模型效果-训练集:
    print "\nModel Report- train data"
    print "Accuracy（Train） : %.4g" % metrics.accuracy_score(dtrain['target'].values, dtrain_predictions)
    print "AUC Score (Train): %f" % metrics.roc_auc_score(dtrain['target'], dtrain_predprob)
    print "GINI Score (Train): %f" % gini_normalized(dtrain['target'], dtrain_predprob)
    
    #模型预测-测试集
    dtest['prediction'] = alg.predict(dtest[predictors])
    dtest['predprob'] = alg.predict_proba(dtest[predictors])[:,1]
    #输出模型效果-测试集
    print "\nModel Report- test data"
    print "Accuracy（Test） : %.4g" % metrics.accuracy_score(dtest['target'].values,dtest['prediction'])
    print 'AUC Score (Test): %f' % metrics.roc_auc_score(dtest['target'], dtest['predprob'])
    print "GINI Score (Test): %f" % gini_normalized(dtest['target'], dtest['predprob'])
    
    return alg

In [9]:
predictors1 = [x for x in train_old.columns if x not in [target, IDcol,tag]]

xgb1 = XGBClassifier(
        learning_rate =0.1,
        n_estimators=20,
        max_depth=5,
        min_child_weight=1,
        gamma=0,
        subsample=0.8,
        colsample_bytree=0.8,
        objective= 'binary:logistic',
        nthread=4,
        scale_pos_weight=26.4, #解决样本不均衡的问题
        seed=27)

alg_before=xgboostModelFit(xgb1, train1, validation_old, predictors1)


Model Report- train data
Accuracy（Train） : 0.6274
AUC Score (Train): 0.644518
GINI Score (Train): 0.289036

Model Report- test data
Accuracy（Test） : 0.6281
AUC Score (Test): 0.644204
GINI Score (Test): 0.288419


In [10]:
predictors2 = [x for x in train_extends.columns if x not in [target, IDcol,tag]]

xgb2 = XGBClassifier(
        learning_rate =0.1,
        n_estimators=20,
        max_depth=5,
        min_child_weight=1,
        gamma=0,
        subsample=0.8,
        colsample_bytree=0.8,
        objective= 'binary:logistic',
        nthread=4,
        scale_pos_weight=26.4, #解决样本不均衡的问题
        seed=27)

alg_after=xgboostModelFit(xgb2, train_extends, validation_extends, predictors2)


Model Report- train data
Accuracy（Train） : 0.7402
AUC Score (Train): 0.795844
GINI Score (Train): 0.592536

Model Report- test data
Accuracy（Test） : 0.7401
AUC Score (Test): 0.789326
GINI Score (Test): 0.579483


# 结论

模型及参数相同的情况下，增加特征工程后，对最终结果有正向效果。

特征工程之前： GINI Score (Train)为 0.289036

特征工程之后： GINI Score (Train)为 0.589044

不过，新训练集的GINI系数偏高，存在过拟合的问题。

# 以下是最终测试数据效果

## 特征工程之前

In [11]:
test1 = pd.read_csv("/Users/leidelong/competition/Porto_Seguro_Safe_Driver_Prediction/input/test.csv")


In [12]:
#最终测试集
test_raw= test1
X_test_raw= test1[predictors1]
y_raw_test_pred = alg_before.predict(X_test_raw)
y_raw_test_predprob=alg_before.predict_proba(X_test_raw)[:,1]

In [13]:
X_ID = test_raw['id']
X_ID=pd.DataFrame(X_ID)
df_y_test_predprob=pd.DataFrame(y_raw_test_predprob)
print df_y_test_predprob.shape,X_ID.shape

(892816, 1) (892816, 1)


In [14]:
result=pd.concat([X_ID,df_y_test_predprob],axis=1)
result.columns = ['id', 'target']
result.shape

(892816, 2)

In [15]:
result.to_csv("/Users/leidelong/competition/Porto_Seguro_Safe_Driver_Prediction/test_result_old.csv",index=False)

## 特征工程之后

In [16]:
test2 = pd.read_csv("/Users/leidelong/competition/Porto_Seguro_Safe_Driver_Prediction/input/test_extends.csv")

In [17]:
#最终测试集
test_raw= test2
X_test_raw= test2[predictors2]
y_raw_test_pred = alg_after.predict(X_test_raw)
y_raw_test_predprob=alg_after.predict_proba(X_test_raw)[:,1]

In [18]:
X_ID = test_raw['id']
X_ID=pd.DataFrame(X_ID)
df_y_test_predprob=pd.DataFrame(y_raw_test_predprob)
print df_y_test_predprob.shape,X_ID.shape


(892816, 1) (892816, 1)


In [19]:
result=pd.concat([X_ID,df_y_test_predprob],axis=1)
result.columns = ['id', 'target']
result.shape

(892816, 2)

In [20]:
result.to_csv("/Users/leidelong/competition/Porto_Seguro_Safe_Driver_Prediction/test_result.csv",index=False)

# LB

在leaderboard上，特征工程之前0.260，特征工程之后0.155。初步认为是特征工程中存在问题。